polar-core/src/debugger.rs

```diff
@@ -172,23 +111,51 @@ impl Debugger {
     /// - `Some(Goal::Debug { message })` -> Pause evaluation.
     /// - `None` -> Continue evaluation.
     fn maybe_break(&self, event: DebugEvent, vm: &PolarVirtualMachine) -> Option<Rc<Goal>> {
-        self.step.as_ref().and_then(|step| match (step, event) {
-            (Step::Goal, DebugEvent::Goal(goal)) => Some(Rc::new(Goal::Debug {
-                message: goal.to_string(),
-            })),
-            (Step::Over { snapshot }, DebugEvent::Query)
-            | (Step::Out { snapshot }, DebugEvent::Query)
-                if vm.queries[..vm.queries.len() - 1] == snapshot[..] =>
+        if let Some(step) = self.step.as_ref() {
+            match (step, event) {
+                (Step::Goal, DebugEvent::Goal(goal)) => Some(Rc::new(Goal::Debug {
+                    message: goal.to_string(),
+                })),
+                (Step::Into, DebugEvent::Query) => self.break_query(vm),
+                (Step::Out { level }, DebugEvent::Query)
+                    if vm.trace_stack.is_empty() || vm.trace_stack.len() < *level =>
+                {
+                    self.break_query(vm)
+                }
+                (Step::Over { level }, DebugEvent::Query) if vm.trace_stack.len() == *level => {
+                    self.break_query(vm)
+                }
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Produce the `Goal::Debug` for breaking on a Query (as opposed to breaking on a Goal).
+    /// This is used to implement the `step`, `over`, and `out` debug commands.
+    pub fn break_query(&self, vm: &PolarVirtualMachine) -> Option<Rc<Goal>> {
+        let message = vm.trace.last().and_then(|trace| {
+            if let Trace {
+                node: Node::Term(q),
+                ..
+            } = &**trace
             {
-                Some(Rc::new(Goal::Debug {
-                    message: vm.queries.last().map_or_else(
-                        || "".to_string(),
-                        |query| self.query_source(&query, &vm.kb.read().unwrap().sources, 0),
-                    ),
-                }))
+                match q.value() {
+                    Value::Expression(Operation {
+                        operator: Operator::And,
+                        args,
+                    }) if args.len() == 1 => None,
+                    _ => {
```

Comment:
nit: Elsewhere in this file, we use a `matches!` macro for this same check. Not sure which I prefer; just pointing out the dissonance.

```rust
if matches!(t.value(), Value::Expression(Operation { operator: Operator::And, args}) if args.len() == 1)
```